# Final Project


In [13]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Project Description

The following project ...


## Data Import

For this project, we will mainly be using two data files: `fma-rock-vs-hiphop.csv` and `echonest-metrics.json`.


In [14]:
df_csv = pd.read_csv("data/fma-rock-vs-hiphop.csv")
df_json = pd.read_json("data/echonest-metrics.json", precise_float=True)

### CSV File

The `fma-rock-vs-hiphop.csv` file contains some general information for songs, having 17,734 songs and 21 variables.


In [15]:
df_csv.head()

track_id  bit_rate  comments        composer         date_created  \
0       135    256000         1             NaN  2008-11-26 01:43:26   
1       136    256000         1             NaN  2008-11-26 01:43:35   
2       151    192000         0             NaN  2008-11-26 01:44:55   
3       152    192000         0             NaN  2008-11-26 01:44:58   
4       153    256000         0  Arc and Sender  2008-11-26 01:45:00   

         date_recorded  duration  favorites genre_top    genres  ...  \
0  2008-11-26 00:00:00       837          0      Rock  [45, 58]  ...   
1  2008-11-26 00:00:00       509          0      Rock  [45, 58]  ...   
2                  NaN       192          0      Rock      [25]  ...   
3                  NaN       193          0      Rock      [25]  ...   
4  2008-11-26 00:00:00       405          5      Rock      [26]  ...   

  information interest  language_code  \
0         NaN     2484             en   
1         NaN     1948             en   
2         NaN      701             en   
3         NaN      637             en   
4         NaN      354             en   

                                             license listens  lyricist number  \
0  Attribution-NonCommercial-ShareAlike 3.0 Inter...    1832       NaN      0   
1  Attribution-NonCommercial-ShareAlike 3.0 Inter...    1498       NaN      0   
2  Attribution-NonCommercial-ShareAlike 3.0 Inter...     148       NaN      4   
3  Attribution-NonCommercial-ShareAlike 3.0 Inter...      98       NaN     11   
4  Attribution-NonCommercial-NoDerivatives (aka M...     424       NaN      2   

   publisher tags                       title  
0        NaN   []                Father's Day  
1        NaN   []  Peel Back The Mountain Sky  
2        NaN   []                 Untitled 04  
3        NaN   []                 Untitled 11  
4        NaN   []          Hundred-Year Flood  

[5 rows x 21 columns]

In [16]:
df_csv.shape

(17734, 21)

### JSON File

On the other hand, `echonest-metrics.json` file contains song metadata for 13,129 songs across 9 variables.


In [17]:
df_json.head()

track_id  acousticness  danceability    energy  instrumentalness  liveness  \
0         2      0.416675      0.675894  0.634476          0.010628  0.177647   
1         3      0.374408      0.528643  0.817461          0.001851  0.105880   
2         5      0.043567      0.745566  0.701470          0.000697  0.373143   
3        10      0.951670      0.658179  0.924525          0.965427  0.115474   
4       134      0.452217      0.513238  0.560410          0.019443  0.096567   

   speechiness    tempo   valence  
0     0.159310  165.922  0.576661  
1     0.461818  126.957  0.269240  
2     0.124595  100.260  0.621661  
3     0.032985  111.562  0.963590  
4     0.525519  114.290  0.894072

In [18]:
df_json.shape

(13129, 9)

After observing both DataFrames we discover that they both share the `track_id` attribute, which can be used in order to combine them.


## DataFrame Combination

After some research, we found that Pandas has 3 different functions used to combine DataFrames:

- `merge()`
- `concat()`
- `join()`

All of them will be tested to find the most optimal one.


As an initial preventive measure, we can perform a test to find how many common `track_id` values are found along both DataFrames.


In [21]:
count = 0
for id in df_json.track_id:
    if id in df_csv.track_id.values:
        count = count + 1

print(count)

4802


This means that any combination of the .csv and the .json DataFrames should have 4802 rows, plus 30 attribute columns.


## Merge


In [ ]:
df_merge = pd.merge(df_csv, df_json, on="track_id")
df_merge.head()

track_id  bit_rate  comments        composer         date_created  \
0       153    256000         0  Arc and Sender  2008-11-26 01:45:00   
1       154    256000         0  Arc and Sender  2008-11-26 01:45:04   
2       155    192000         0             NaN  2008-11-26 01:45:10   
3       169    192000         0   James Squeaky  2008-11-26 01:46:02   
4       170    192000         0             NaN  2008-11-26 01:46:05   

         date_recorded  duration  favorites genre_top genres  ... tags  \
0  2008-11-26 00:00:00       405          5      Rock   [26]  ...   []   
1  2008-11-26 00:00:00       319          1      Rock   [26]  ...   []   
2  2008-11-26 00:00:00       756          1      Rock   [26]  ...   []   
3  2006-01-01 00:00:00       144          1      Rock   [25]  ...   []   
4  2006-10-07 00:00:00       181          0      Rock   [25]  ...   []   

                        title  acousticness danceability    energy  \
0          Hundred-Year Flood      0.988306     0.255661  0.979774   
1         Squares And Circles      0.970135     0.352946  0.023852   
2     Maps of the Stars Homes      0.981657     0.142249  0.912122   
3                Boss of Goth      0.989141     0.225978  0.722835   
4  Industry Standard Massacre      0.886660     0.298518  0.744333   

   instrumentalness  liveness  speechiness   tempo   valence  
0          0.973006  0.121342     0.051740  90.241  0.034018  
1          0.957113  0.113261     0.032177  53.758  0.035632  
2          0.967294  0.363510     0.087527  91.912  0.034325  
3          0.263076  0.092371     0.053406  94.322  0.028347  
4          0.920950  0.139587     0.088781  97.880  0.073548  

[5 rows x 29 columns]

In [ ]:
df_merge.shape

(4802, 29)